// skimmed-datepicker/esm5/lib/view-strategies/days-strategy.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive } from '@angular/core';
import { ViewStrategy } from './view-strategy';
import { ViewMode } from '../datepicker/view-mode';
import { startOfMonth } from '../util/helpers';
var DaysStrategyDirective = /** @class */ (function (_super) {
    tslib_1.__extends(DaysStrategyDirective, _super);
    function DaysStrategyDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.viewMode = ViewMode.Days;
        _this.itemClass = 'skm-datepicker-day';
        return _this;
    }
    /**
     * @param {?} timestamp
     * @return {?}
     */
    DaysStrategyDirective.prototype.normalizeTimestamp = /**
     * @param {?} timestamp
     * @return {?}
     */
    function (timestamp) {
        return timestamp;
    };
    /**
     * @param {?} timestamp
     * @return {?}
     */
    DaysStrategyDirective.prototype.makeInitPanesSeed = /**
     * @param {?} timestamp
     * @return {?}
     */
    function (timestamp) {
        return startOfMonth(timestamp).valueOf();
    };
    /**
     * @param {?} timestamp
     * @param {?} add
     * @param {?} baseOrder
     * @param {?} weekStart
     * @return {?}
     */
    DaysStrategyDirective.prototype.makePane = /**
     * @param {?} timestamp
     * @param {?} add
     * @param {?} baseOrder
     * @param {?} weekStart
     * @return {?}
     */
    function (timestamp, add, baseOrder, weekStart) {
        var /** @type {?} */ date = new Date(timestamp);
        date.setMonth(add + date.getMonth());
        var /** @type {?} */ firstDay = date.getDay();
        date.setMonth(1 + date.getMonth());
        date.setDate(0);
        var /** @type {?} */ monthLength = date.getDate();
        var /** @type {?} */ values = [];
        for (var /** @type {?} */ i = 1; i <= monthLength; i++) {
            values.push(date.setDate(i));
        }
        return {
            order: baseOrder + add,
            values: values,
            indent: (firstDay - weekStart + 7) % 7 || 7,
        };
    };
    DaysStrategyDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[skmDaysStrategy]',
                    providers: [
                        { provide: ViewStrategy, useExisting: DaysStrategyDirective },
                    ],
                },] },
    ];
    return DaysStrategyDirective;
}(ViewStrategy));
export { DaysStrategyDirective };
function DaysStrategyDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    DaysStrategyDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    DaysStrategyDirective.ctorParameters;
    /** @type {?} */
    DaysStrategyDirective.prototype.viewMode;
    /** @type {?} */
    DaysStrategyDirective.prototype.itemClass;
}
