// skimmed-datepicker/esm5/lib/view-strategies/months-strategy.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive } from '@angular/core';
import { startOfYear } from '../util/helpers';
import { ViewStrategy } from './view-strategy';
import { ViewMode } from '../datepicker/view-mode';
var MonthsStrategyDirective = /** @class */ (function (_super) {
    tslib_1.__extends(MonthsStrategyDirective, _super);
    function MonthsStrategyDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.viewMode = ViewMode.Months;
        _this.itemClass = 'skm-datepicker-month';
        return _this;
    }
    /**
     * @param {?} timestamp
     * @return {?}
     */
    MonthsStrategyDirective.prototype.normalizeTimestamp = /**
     * @param {?} timestamp
     * @return {?}
     */
    function (timestamp) {
        return new Date(timestamp).setDate(1);
    };
    /**
     * @param {?} timestamp
     * @return {?}
     */
    MonthsStrategyDirective.prototype.makeInitPanesSeed = /**
     * @param {?} timestamp
     * @return {?}
     */
    function (timestamp) {
        return startOfYear(timestamp).valueOf();
    };
    /**
     * @param {?} timestamp
     * @param {?} add
     * @param {?} baseOrder
     * @param {?} weekStart
     * @return {?}
     */
    MonthsStrategyDirective.prototype.makePane = /**
     * @param {?} timestamp
     * @param {?} add
     * @param {?} baseOrder
     * @param {?} weekStart
     * @return {?}
     */
    function (timestamp, add, baseOrder, weekStart) {
        var /** @type {?} */ date = new Date(timestamp);
        date.setFullYear(add + date.getFullYear());
        var /** @type {?} */ values = [];
        for (var /** @type {?} */ i = 0; i < 12; i++) {
            values.push(date.setMonth(i));
        }
        return {
            order: baseOrder + add,
            values: values,
        };
    };
    MonthsStrategyDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[skmMonthsStrategy]',
                    providers: [
                        { provide: ViewStrategy, useExisting: MonthsStrategyDirective },
                    ],
                },] },
    ];
    return MonthsStrategyDirective;
}(ViewStrategy));
export { MonthsStrategyDirective };
function MonthsStrategyDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MonthsStrategyDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MonthsStrategyDirective.ctorParameters;
    /** @type {?} */
    MonthsStrategyDirective.prototype.viewMode;
    /** @type {?} */
    MonthsStrategyDirective.prototype.itemClass;
}
