// skimmed-datepicker/esm5/lib/view-strategies/years-strategy.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive } from '@angular/core';
import { startOfYear } from '../util/helpers';
import { ViewStrategy } from './view-strategy';
import { ViewMode } from '../datepicker/view-mode';
var YearsStrategyDirective = /** @class */ (function (_super) {
    tslib_1.__extends(YearsStrategyDirective, _super);
    function YearsStrategyDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.viewMode = ViewMode.Years;
        _this.itemClass = 'skm-datepicker-year';
        return _this;
    }
    /**
     * @param {?} timestamp
     * @return {?}
     */
    YearsStrategyDirective.prototype.normalizeTimestamp = /**
     * @param {?} timestamp
     * @return {?}
     */
    function (timestamp) {
        return new Date(timestamp).setMonth(0, 1);
    };
    /**
     * @param {?} timestamp
     * @return {?}
     */
    YearsStrategyDirective.prototype.makeInitPanesSeed = /**
     * @param {?} timestamp
     * @return {?}
     */
    function (timestamp) {
        var /** @type {?} */ date = new Date(timestamp);
        var /** @type {?} */ origin = date.getFullYear();
        var /** @type {?} */ adjusted = origin - (origin % 20);
        return startOfYear(date).setFullYear(adjusted);
    };
    /**
     * @param {?} timestamp
     * @param {?} add
     * @param {?} baseOrder
     * @param {?} weekStart
     * @return {?}
     */
    YearsStrategyDirective.prototype.makePane = /**
     * @param {?} timestamp
     * @param {?} add
     * @param {?} baseOrder
     * @param {?} weekStart
     * @return {?}
     */
    function (timestamp, add, baseOrder, weekStart) {
        var /** @type {?} */ date = new Date(timestamp);
        var /** @type {?} */ origin = add * 20 + date.getFullYear();
        var /** @type {?} */ values = [];
        for (var /** @type {?} */ i = 0; i < 20; i++) {
            values.push(date.setFullYear(origin + i));
        }
        return {
            order: baseOrder + add,
            values: values,
        };
    };
    YearsStrategyDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[skmYearsStrategy]',
                    providers: [
                        { provide: ViewStrategy, useExisting: YearsStrategyDirective },
                    ],
                },] },
    ];
    return YearsStrategyDirective;
}(ViewStrategy));
export { YearsStrategyDirective };
function YearsStrategyDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    YearsStrategyDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    YearsStrategyDirective.ctorParameters;
    /** @type {?} */
    YearsStrategyDirective.prototype.viewMode;
    /** @type {?} */
    YearsStrategyDirective.prototype.itemClass;
}
