// skimmed-datepicker/esm2015/lib/view-strategies/days-strategy.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive } from '@angular/core';
import { ViewStrategy } from './view-strategy';
import { ViewMode } from '../datepicker/view-mode';
import { startOfMonth } from '../util/helpers';
export class DaysStrategyDirective extends ViewStrategy {
    constructor() {
        super(...arguments);
        this.viewMode = ViewMode.Days;
        this.itemClass = 'skm-datepicker-day';
    }
    /**
     * @param {?} timestamp
     * @return {?}
     */
    normalizeTimestamp(timestamp) {
        return timestamp;
    }
    /**
     * @param {?} timestamp
     * @return {?}
     */
    makeInitPanesSeed(timestamp) {
        return startOfMonth(timestamp).valueOf();
    }
    /**
     * @param {?} timestamp
     * @param {?} add
     * @param {?} baseOrder
     * @param {?} weekStart
     * @return {?}
     */
    makePane(timestamp, add, baseOrder, weekStart) {
        const /** @type {?} */ date = new Date(timestamp);
        date.setMonth(add + date.getMonth());
        const /** @type {?} */ firstDay = date.getDay();
        date.setMonth(1 + date.getMonth());
        date.setDate(0);
        const /** @type {?} */ monthLength = date.getDate();
        const /** @type {?} */ values = [];
        for (let /** @type {?} */ i = 1; i <= monthLength; i++) {
            values.push(date.setDate(i));
        }
        return {
            order: baseOrder + add,
            values: values,
            indent: (firstDay - weekStart + 7) % 7 || 7,
        };
    }
}
DaysStrategyDirective.decorators = [
    { type: Directive, args: [{
                selector: '[skmDaysStrategy]',
                providers: [
                    { provide: ViewStrategy, useExisting: DaysStrategyDirective },
                ],
            },] },
];
function DaysStrategyDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    DaysStrategyDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    DaysStrategyDirective.ctorParameters;
    /** @type {?} */
    DaysStrategyDirective.prototype.viewMode;
    /** @type {?} */
    DaysStrategyDirective.prototype.itemClass;
}
