// skimmed-datepicker/esm2015/lib/view-strategies/months-strategy.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive } from '@angular/core';
import { startOfYear } from '../util/helpers';
import { ViewStrategy } from './view-strategy';
import { ViewMode } from '../datepicker/view-mode';
export class MonthsStrategyDirective extends ViewStrategy {
    constructor() {
        super(...arguments);
        this.viewMode = ViewMode.Months;
        this.itemClass = 'skm-datepicker-month';
    }
    /**
     * @param {?} timestamp
     * @return {?}
     */
    normalizeTimestamp(timestamp) {
        return new Date(timestamp).setDate(1);
    }
    /**
     * @param {?} timestamp
     * @return {?}
     */
    makeInitPanesSeed(timestamp) {
        return startOfYear(timestamp).valueOf();
    }
    /**
     * @param {?} timestamp
     * @param {?} add
     * @param {?} baseOrder
     * @param {?} weekStart
     * @return {?}
     */
    makePane(timestamp, add, baseOrder, weekStart) {
        const /** @type {?} */ date = new Date(timestamp);
        date.setFullYear(add + date.getFullYear());
        const /** @type {?} */ values = [];
        for (let /** @type {?} */ i = 0; i < 12; i++) {
            values.push(date.setMonth(i));
        }
        return {
            order: baseOrder + add,
            values: values,
        };
    }
}
MonthsStrategyDirective.decorators = [
    { type: Directive, args: [{
                selector: '[skmMonthsStrategy]',
                providers: [
                    { provide: ViewStrategy, useExisting: MonthsStrategyDirective },
                ],
            },] },
];
function MonthsStrategyDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MonthsStrategyDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MonthsStrategyDirective.ctorParameters;
    /** @type {?} */
    MonthsStrategyDirective.prototype.viewMode;
    /** @type {?} */
    MonthsStrategyDirective.prototype.itemClass;
}
