// skimmed-datepicker/esm2015/lib/view-strategies/years-strategy.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive } from '@angular/core';
import { startOfYear } from '../util/helpers';
import { ViewStrategy } from './view-strategy';
import { ViewMode } from '../datepicker/view-mode';
export class YearsStrategyDirective extends ViewStrategy {
    constructor() {
        super(...arguments);
        this.viewMode = ViewMode.Years;
        this.itemClass = 'skm-datepicker-year';
    }
    /**
     * @param {?} timestamp
     * @return {?}
     */
    normalizeTimestamp(timestamp) {
        return new Date(timestamp).setMonth(0, 1);
    }
    /**
     * @param {?} timestamp
     * @return {?}
     */
    makeInitPanesSeed(timestamp) {
        const /** @type {?} */ date = new Date(timestamp);
        const /** @type {?} */ origin = date.getFullYear();
        const /** @type {?} */ adjusted = origin - (origin % 20);
        return startOfYear(date).setFullYear(adjusted);
    }
    /**
     * @param {?} timestamp
     * @param {?} add
     * @param {?} baseOrder
     * @param {?} weekStart
     * @return {?}
     */
    makePane(timestamp, add, baseOrder, weekStart) {
        const /** @type {?} */ date = new Date(timestamp);
        const /** @type {?} */ origin = add * 20 + date.getFullYear();
        const /** @type {?} */ values = [];
        for (let /** @type {?} */ i = 0; i < 20; i++) {
            values.push(date.setFullYear(origin + i));
        }
        return {
            order: baseOrder + add,
            values: values,
        };
    }
}
YearsStrategyDirective.decorators = [
    { type: Directive, args: [{
                selector: '[skmYearsStrategy]',
                providers: [
                    { provide: ViewStrategy, useExisting: YearsStrategyDirective },
                ],
            },] },
];
function YearsStrategyDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    YearsStrategyDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    YearsStrategyDirective.ctorParameters;
    /** @type {?} */
    YearsStrategyDirective.prototype.viewMode;
    /** @type {?} */
    YearsStrategyDirective.prototype.itemClass;
}
