// skimmed-datepicker/esm5/lib/grid/grid.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding, Input } from '@angular/core';
import { SliderComponent } from '../slider/slider.component';
import { ViewStrategy } from '../view-strategies/view-strategy';
import { ViewComponent } from '../view/view.component';
var GridComponent = /** @class */ (function () {
    function GridComponent(slider, parentView, viewStrategy) {
        this.slider = slider;
        this.parentView = parentView;
        this.viewStrategy = viewStrategy;
        this.itemLabels = [];
        this._hostClass = 'skm-datepicker-content';
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    GridComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if ('itemLabels' in changes) {
            this.itemLabels = this.itemLabels || [];
        }
    };
    /**
     * @param {?} index
     * @param {?} pane
     * @return {?}
     */
    GridComponent.prototype.makeItemClasses = /**
     * @param {?} index
     * @param {?} pane
     * @return {?}
     */
    function (index, pane) {
        return [
            this.viewStrategy.itemClass,
            'skm-datepicker-item',
            (pane.values[index] === this.currentTimestamp) ? 'skm-datepicker-current' : '',
            (pane.values[index] === this.selectedTimestamp) ? 'skm-datepicker-selected' : '',
        ].join(' ');
    };
    /**
     * @param {?} index
     * @param {?} pane
     * @return {?}
     */
    GridComponent.prototype.isDisabled = /**
     * @param {?} index
     * @param {?} pane
     * @return {?}
     */
    function (index, pane) {
        return (pane.values[index] < this.minTimestamp) || (pane.values[index] > this.maxTimestamp);
    };
    /**
     * @param {?} index
     * @return {?}
     */
    GridComponent.prototype.trackIndex = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        return index;
    };
    GridComponent.decorators = [
        { type: Component, args: [{
                    selector: 'skm-grid',
                    template: "<div class=\"skm-datepicker-grid\" (click)=\"parentView.selectItem($event, pane, slider.notPanning)\">\n  <div class=\"skm-datepicker-indent\" [style.width.%]=\"100 / 7 * pane.indent\"></div>\n  <div *ngFor=\"let value of pane.values; trackBy: trackIndex; let i = index\" class=\"skm-datepicker-cell\">\n    <button type=\"button\" [attr.data-index]=\"i\" [class]=\"makeItemClasses(i, pane)\" [disabled]=\"isDisabled(i, pane)\">\n      {{itemLabels[i] || (value | date:itemFormat)}}\n    </button>\n  </div>\n</div>\n",
                    styles: ["button{display:block;color:inherit}button:not([disabled]){cursor:pointer}button[disabled]{pointer-events:none}.skm-datepicker-pane{flex-shrink:0;width:100%;display:flex;flex-direction:column;overflow:hidden}.skm-datepicker-week{display:flex}.skm-datepicker-weekday{width:14.28571%}.skm-datepicker-content{flex:1 0 0;display:flex;flex-direction:column}.skm-datepicker-grid{flex:1 0 0;display:flex;flex-wrap:wrap;align-content:flex-start;pointer-events:none}.skm-datepicker-indent{height:16.66667%}.skm-datepicker-cell{display:flex}:host-context(.skm-datepicker-days-view) .skm-datepicker-cell{width:14.28571%;height:16.66667%}:host-context(.skm-datepicker-months-view) .skm-datepicker-cell{width:33.33333%;height:25%}:host-context(.skm-datepicker-years-view) .skm-datepicker-cell{width:20%;height:25%}.skm-datepicker-item{flex:1;pointer-events:auto}"],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                },] },
    ];
    /** @nocollapse */
    GridComponent.ctorParameters = function () { return [
        { type: SliderComponent, },
        { type: ViewComponent, },
        { type: ViewStrategy, },
    ]; };
    GridComponent.propDecorators = {
        "pane": [{ type: Input },],
        "currentTimestamp": [{ type: Input },],
        "selectedTimestamp": [{ type: Input },],
        "minTimestamp": [{ type: Input },],
        "maxTimestamp": [{ type: Input },],
        "itemFormat": [{ type: Input },],
        "itemLabels": [{ type: Input },],
        "_hostClass": [{ type: HostBinding, args: ['class',] },],
    };
    return GridComponent;
}());
export { GridComponent };
function GridComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    GridComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    GridComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    GridComponent.propDecorators;
    /** @type {?} */
    GridComponent.prototype.pane;
    /** @type {?} */
    GridComponent.prototype.currentTimestamp;
    /** @type {?} */
    GridComponent.prototype.selectedTimestamp;
    /** @type {?} */
    GridComponent.prototype.minTimestamp;
    /** @type {?} */
    GridComponent.prototype.maxTimestamp;
    /** @type {?} */
    GridComponent.prototype.itemFormat;
    /** @type {?} */
    GridComponent.prototype.itemLabels;
    /** @type {?} */
    GridComponent.prototype._hostClass;
    /** @type {?} */
    GridComponent.prototype.slider;
    /** @type {?} */
    GridComponent.prototype.parentView;
    /** @type {?} */
    GridComponent.prototype.viewStrategy;
}
